### PR TITLE
Doc: ptref grammar

### PR DESCRIPTION
**Pull request by TeXitoi**
Rendered version: https://github.com/TeXitoi/navitia/blob/document_ptref_grammar/documentation/rfc/ptref_grammar.md

### CODE REVIEW

**Comment by TeXitoi** [comment]
@eturck @vpassama can you review this?

File: documentation/rfc/ptref_grammar.md [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+Examples:
+ * `/v1/coverage/id/stop_areas?filter=all`: returns all the stop areas of the coverage.
+ * `/v1/coverage/id/stop_areas?filter=none`: returns 0 stop areas (you get a 404).
+ 
+### Method predicates
+
+Method predicates are in the form `collection.method(string, string, ...)` where:
+ * `collection` is a collection name (as the collections queried, but singular);
+ * `method` is a method name, corresponding to the filter on the `collection`;
+ * arguments to the method between `(` and `)`, as strings, separated by `,`.
```
Comment (antoine-de) [inline documentation/rfc/ptref_grammar.md:34]: an example would be nice

**Review by antoine-de — APPROVED** [review]
great explanations!

(and good work on pt_ref next gen)

File: documentation/rfc/ptref_grammar.md (hunk 1 above)
Comment (TeXitoi) [reply documentation/rfc/ptref_grammar.md:34]: Which kind of example? There is the list of methods just after, with examples.

Comment (antoine-de) [reply documentation/rfc/ptref_grammar.md:34]: yes the later examples are really great, but well, they are ... later 😉 so I read them after.

maybe just something like:
`stop_area.has_code("neptune","1242")` ? or any code that might be more realistic

**Comment by eturck** [comment]
That's really great. Thank you for documenting it so clearly.

**Review by eturck — APPROVED** [review]

File: documentation/rfc/ptref_grammar.md (hunk 1 above)
Comment (TeXitoi) [reply documentation/rfc/ptref_grammar.md:34]: Better like that?

File: documentation/rfc/ptref_grammar.md [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+
+The `arrivals`, `departures` and `stop_schedules` endpoint builds their responses on the `journey_pattern_points` collection.
+
+## Predicates
+
+The elementary building block of a query is the predicate. They define a set of requested objects.
+
+### Elementary predicates
+
+There is 2 elementary predicates: `all` and `empty`. As expected, `all` returns all the queried objects while `empty` returns none.
```
Comment (nonifier) [inline documentation/rfc/ptref_grammar.md]:
```suggestion
There are 2 elementary predicates: `all` and `empty`. As expected, `all` returns all the queried objects while `empty` returns none.
```

File: documentation/rfc/ptref_grammar.md [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+ * `1337`
+ * `-42`
+ * `-123.45e-12`
+ * `2.37715;48.846781`
+ * `"I can have spaces and accentuated letters: éèà"` (decoded as `I can have spaces and accentuated letters: éèà`)
+ * `"a \" and a \\ within a string"` (decoded as `a " and a \ within a string`)
+
+#### Availlable methods
+
+In the following table, if the invocation is begining with `collection`, any collection can be used, else the method will only work for the specified collection.
```
Comment (nonifier) [inline documentation/rfc/ptref_grammar.md]:
```suggestion
In the following table, if the invocation starts with `collection`, any collection can be used, else the method will only work for the specified collection.
```

File: documentation/rfc/ptref_grammar.md [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+|`pred1 or get collection <- pred2 - pred3`|`pred1 or (get collection <- (pred2 - pred3))`|
+
+
+Examples:
+* `/v1/coverage/id/lines?filter=get network <- line.id=Metro14` returns the lines from the network of the line Metro14.
+* `/v1/coverage/id/physical_modes?filter=get connection <- line.id=Metro14` returns the physical modes in connection with the line Metro14.
+
+## Advanced examples
+
+### Get the list of stop areas you can go from a given stop area without connection using a given network
```
Comment (nonifier) [inline documentation/rfc/ptref_grammar.md]:
```suggestion
### Get the list of stop areas accessible from a given stop area, a network and without connection.
```

File: documentation/rfc/ptref_grammar.md [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+
+You are doing a time table between 2 stop areas for a specific network. In the interface, you first select the first stop area, and then, you must propose the candidate second stop areas.
```
Comment (nonifier) [inline documentation/rfc/ptref_grammar.md]: Not sure I understand that comment. I don't see how this relates to the query :thinking: ?

**Review by nonifier — APPROVED** [review]
That's a bloody good doc !! :1st_place_medal:

File: documentation/rfc/ptref_grammar.md [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+
+> `/v1/coverage/id/stop_areas?filter=get vehicle_journey <- network.id=MyNetwork and stop_area.id=FirstStopArea`
+
+From this, we now need to remove the first stop area:
+
+> `/v1/coverage/id/stop_areas?filter=(get vehicle_journey <- network.id=MyNetwork and stop_area.id=FirstStopArea) - stop_area.id=FirstStopArea`
+
+### Get all the line that are not composed only of buses in connection with a given line
+
+In the home page of a line, you want to list the lines in connection. But, as there is a lot of bus lines, you only want the lines that contains other modes.
```
Comment (xlqian) [inline documentation/rfc/ptref_grammar.md]:
```suggestion
In the home page of a line, you want to list the lines in connection. But, as there are a lot of bus lines, you only want the lines that contain other modes.
```

**Review by xlqian — APPROVED** [review]
:+1:

File: documentation/rfc/ptref_grammar.md (hunk 5 above)
Comment (TeXitoi) [reply documentation/rfc/ptref_grammar.md]: Is it better now?

File: documentation/rfc/ptref_grammar.md [hunk 7]
```diff
@@ -0,0 +1,225 @@
+# Public Transit Referential grammar
+
+It's possible to do advanced query to the public transit referential. The query is in the form `/v1/coverage/id/collections?filter=query`. This is the documentation of the `query` syntax. For simplicity, the queries in this documentation will not be urlencoded, but it must be in practice.
```
Comment (pbougue) [inline documentation/rfc/ptref_grammar.md]: Minor:
"This is the documentation of the `'collections' and `'query' syntax. `'query' must be urlencoded in practice, but for simplicity it is not the case in this documentation.`"

File: documentation/rfc/ptref_grammar.md [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+
+The `line_reports` endpoint builds its response on the `lines` collection.
+
+The `route_schedules` endpoint builds its response on the `routes` collections.
+
+The `arrivals`, `departures` and `stop_schedules` endpoint builds their responses on the `journey_pattern_points` collection.
+
+## Predicates
+
+The elementary building block of a query is the predicate. They define a set of requested objects.
```
Comment (pbougue) [inline documentation/rfc/ptref_grammar.md]: "`It` define`s` a set of ..."

File: documentation/rfc/ptref_grammar.md [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+ * `collection` is a collection name (as the collections queried, but singular);
+ * `method` is a method name, corresponding to the filter on the `collection`;
+ * arguments to the method between `(` and `)`, as strings, separated by `,`.
+ 
+Example: `/v1/coverage/id/stop_areas?filter=stop_area.has_code("neptune","1242")`
+ 
+#### Strings
+ 
+A string can be:
+ * an escaped string, begining and ending with `"`, with `\x` transformed as `x` within the string
```
Comment (pbougue) [inline documentation/rfc/ptref_grammar.md]: typo: begin`n`ing

File: documentation/rfc/ptref_grammar.md [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+ * a basic string, composed of any letter, digit, `_`, `.`, `:`, `;`, `|` and `-`.
+ 
+Basic strings are great to easily type numbers and navitia coordinates. It can also handle most of the identifiers, but be warned that you will run into troubles if you use the basic string syntax in a programatic way (think of identifiers with spaces or accentuated letters).
```
Comment (pbougue) [inline documentation/rfc/ptref_grammar.md]: typo: program`m`atic

File: documentation/rfc/ptref_grammar.md [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+|`disruption.since(since)`|all the disruptions that are active after the given datetime|`since` must be UTC datetime in the ISO 8601 format (ending with `Z`)|
+|`disruption.tag(tag)`|all the disruptions containing the given tag|equivalent to `disruption.tags(tag)`|
+|`disruption.tags(tag1, tag2, ...)`|all the disruptions containing at least one of the given tags|at least one tag must be given|
+|`disruption.until(until)`|all the disruptions that are active before the given datetime|`until` must be UTC datetime in the ISO 8601 format (ending with `Z`)|
+|`line.code(code)`|all the lines containing the given code|this predicate use the field `line.code`, not `line.codes[]`|
+|`line.odt_level(level)`|all the lines with the given on demand transport property|level can be `scheduled`, `with_stops` or `zonal`|
+|`poi.within(distance, coord)`|all the POIs within `distance` meters from `coord`|`distance` in a positive integer, `coord` is a navitia coord (`lon;lat`)|
+|`stop_area.within(distance, coord)`|all the stop areas within `distance` meters from `coord`|`distance` in a positive integer, `coord` is a navitia coord (`lon;lat`)|
```
Comment (pbougue) [inline documentation/rfc/ptref_grammar.md]: super minor: May be grouped in one row

File: documentation/rfc/ptref_grammar.md [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+|`collection.id(id)`|the object of type `collection` that have `id` as identifier (empty if this identifier is not present)|for types without identifier (as `connection`), it is equivalent to `empty`|
+|`collection.uri(id)`|same as `collection.id(id)`|deprecated|
+|`collection.name(name)`|all the objects of type `collection` that have `name` as name|for types without name (as `connection`), it is equivalent to `empty`|
+|`disruption.between(since, until)`|all the disruptions that are active during the `[since, until]` period|`since` and `until` must be UTC datetime in the ISO 8601 format (ending with `Z`)|
+|`disruption.since(since)`|all the disruptions that are active after the given datetime|`since` must be UTC datetime in the ISO 8601 format (ending with `Z`)|
+|`disruption.tag(tag)`|all the disruptions containing the given tag|equivalent to `disruption.tags(tag)`|
+|`disruption.tags(tag1, tag2, ...)`|all the disruptions containing at least one of the given tags|at least one tag must be given|
+|`disruption.until(until)`|all the disruptions that are active before the given datetime|`until` must be UTC datetime in the ISO 8601 format (ending with `Z`)|
```
Comment (pbougue) [inline documentation/rfc/ptref_grammar.md]: better if the row is just after `between` and `since`

File: documentation/rfc/ptref_grammar.md [hunk 13]
```diff
@@ -0,0 +1,225 @@
...
+|`line.odt_level(level)`|all the lines with the given on demand transport property|level can be `scheduled`, `with_stops` or `zonal`|
+|`poi.within(distance, coord)`|all the POIs within `distance` meters from `coord`|`distance` in a positive integer, `coord` is a navitia coord (`lon;lat`)|
+|`stop_area.within(distance, coord)`|all the stop areas within `distance` meters from `coord`|`distance` in a positive integer, `coord` is a navitia coord (`lon;lat`)|
+|`stop_point.within(distance, coord)`|all the stop points within `distance` meters from `coord`|`distance` in a positive integer, `coord` is a navitia coord (`lon;lat`)|
+|`vehicle_journey.between(since, until)`|all the vehicle journeys that start during the `[since, until]` period|`since` and `until` must be UTC datetime in the ISO 8601 format (ending with `Z`)|
+|`vehicle_journey.has_disruption()`|all the vehicle journeys containing at least a disruption ||
+|`vehicle_journey.has_headsign(headsign)`|all the vehicle journeys containing the given headsign ||
+|`vehicle_journey.since(since)`|all the vehicle journeys that start after the given datetime|`since` must be UTC datetime in the ISO 8601 format (ending with `Z`)|
```
Comment (pbougue) [inline documentation/rfc/ptref_grammar.md]: better if those 2 rows are just after `between`

File: documentation/rfc/ptref_grammar.md [hunk 14]
```diff
@@ -0,0 +1,225 @@
...
+ * `/v1/coverage/id/lines?filter=stop_area.name=Nation and stop_area.name="Charles de Gaulle Etoile" - (stop_area.name=Bastille or stop_area.name=Corvisart)` returns the lines that stop at Nation and Charles de Gaulle Etoile but not at Bastille nor Corvisart.
+
+## Sub request
+
+Method predicates allow you to request for a set of object of a given type, and then getting the corresponding objects of another type. Sub request allow you to chain such a pattern. Imagine you want to list the line of the same network of a given line. First, you want to get the networks corresponding to the given line, and then getting the corresponding lines from these networks.
+
+The syntax is `get collection <- expression`. It is lower priority that the set operators and its associativity is right to left. It can be used as a predicate (`(get collection <- expression) or pred`), and nested several times (`get collection1 <- get collection2 <- expression`).
```
Comment (pbougue) [inline documentation/rfc/ptref_grammar.md]: I guess you meant `Its priority is lower than the set of operators`

File: documentation/rfc/ptref_grammar.md [hunk 15]
```diff
@@ -0,0 +1,225 @@
...
+
+Examples:
+* `/v1/coverage/id/lines?filter=get network <- line.id=Metro14` returns the lines from the network of the line Metro14.
+* `/v1/coverage/id/physical_modes?filter=get connection <- line.id=Metro14` returns the physical modes in connection with the line Metro14.
+
+## Advanced examples
+
+### Get the list of stop areas accessible from a given stop area, a network and without connection.
+
+You are doing an interface to construct a time table of all the trains between 2 chosen stop areas. The user select 2 stop areas, and the interface gives all the trains passing by these 2 stop areas with their times of passage.
```
Comment (pbougue) [inline documentation/rfc/ptref_grammar.md]: the user select`s` 2 stop

File: documentation/rfc/ptref_grammar.md [hunk 16]
```diff
@@ -0,0 +1,225 @@
...
+
+To improve the user experience, you want that, one time the user has chosen the first stop area, the interface propose the list of candidate stop areas for the second field. The goal of this example is to construct the query to find the relevant stop areas.
```
Comment (pbougue) [inline documentation/rfc/ptref_grammar.md]: ~one time~ > `once`

File: documentation/rfc/ptref_grammar.md [hunk 17]
```diff
@@ -0,0 +1,225 @@
...
+
+The search is:
+1. get all the trains that pass by the first stop area
+2. get the stop areas where the vehicle journeys stop
```
Comment (pbougue) [inline documentation/rfc/ptref_grammar.md]: vehicle`_`journeys

File: documentation/rfc/ptref_grammar.md [hunk 18]
```diff
@@ -0,0 +1,225 @@
...
+
+Then, for the second step, we get the stop areas corresponding to the previous request:
+
+> `/v1/coverage/id/stop_areas?filter=get vehicle_journey <- physical_mode.id=physical_mode:Train and stop_area.id=FirstStopArea`
+
+From this, we now need to remove the first stop area:
+
+> `/v1/coverage/id/stop_areas?filter=(get vehicle_journey <- physical_mode.id=physical_mode:Train and stop_area.id=FirstStopArea) - stop_area.id=FirstStopArea`
+
+### Get all the line that are not composed only of buses in connection with a given line
```
Comment (pbougue) [inline documentation/rfc/ptref_grammar.md]: Get all the line`s` that

File: documentation/rfc/ptref_grammar.md [hunk 19]
```diff
@@ -0,0 +1,225 @@
...
+### Get the list of stop areas accessible from a given stop area, a network and without connection.
+
+You are doing an interface to construct a time table of all the trains between 2 chosen stop areas. The user select 2 stop areas, and the interface gives all the trains passing by these 2 stop areas with their times of passage.
+
+To improve the user experience, you want that, one time the user has chosen the first stop area, the interface propose the list of candidate stop areas for the second field. The goal of this example is to construct the query to find the relevant stop areas.
+
+The search is:
+1. get all the trains that pass by the first stop area
+2. get the stop areas where the vehicle journeys stop
+3. remove the first stop area from this list
```
Comment (pbougue) [inline documentation/rfc/ptref_grammar.md:179]: minor: the `initial` stop

File: documentation/rfc/ptref_grammar.md [hunk 20]
```diff
@@ -0,0 +1,225 @@
...
+> `/v1/coverage/id/stop_areas?filter=(get vehicle_journey <- physical_mode.id=physical_mode:Train and stop_area.id=FirstStopArea) - stop_area.id=FirstStopArea`
+
+### Get all the line that are not composed only of buses in connection with a given line
+
+In the home page of a line, you want to list the lines in connection. But, as there are a lot of bus lines, you only want the lines that contain other modes.
+
+The search is:
+1. get all the connection of the line
+2. get all the lines from these connections
+3. remove the line
```
Comment (pbougue) [inline documentation/rfc/ptref_grammar.md]: remove the `initial` line

**Review by pbougue — COMMENTED** [review]
Lots of picky :gb: remarks :smile: 
I knew that already, but I kept thinking this is genius, and the doc to explain it is great, clear, progressive :tada: 
:+1: on how you build a complex request and try along the way.

File: documentation/rfc/ptref_grammar.md (hunk 12 above)
Comment (TeXitoi) [reply documentation/rfc/ptref_grammar.md]: That's in alphabetic order

File: documentation/rfc/ptref_grammar.md (hunk 14 above)
Comment (TeXitoi) [reply documentation/rfc/ptref_grammar.md]: No I mean operators on set, but you're right that's just disturbing. Will remove the set word.

File: documentation/rfc/ptref_grammar.md (hunk 17 above)
Comment (TeXitoi) [reply documentation/rfc/ptref_grammar.md]: All the objects in sentences are in English and not code. There are a lot of "stop areas" without underscore.

File: documentation/rfc/ptref_grammar.md (hunk 19 above)
Comment (TeXitoi) [reply documentation/rfc/ptref_grammar.md:179]: That's the first stop in opposition to the second stop of the interface.

File: documentation/rfc/ptref_grammar.md (hunk 14 above)
Comment (pbougue) [reply documentation/rfc/ptref_grammar.md]: Ah, OK, so change the tha`n` too ;)

File: documentation/rfc/ptref_grammar.md (hunk 12 above)
Comment (pbougue) [reply documentation/rfc/ptref_grammar.md]: Not sure what's the most important, anyway it's fine :)

File: documentation/rfc/ptref_grammar.md (hunk 17 above)
Comment (pbougue) [reply documentation/rfc/ptref_grammar.md]: Then it's vehicle's journeys (or trips?). Well, that's minor.

File: documentation/rfc/ptref_grammar.md (hunk 12 above)
Comment (TeXitoi) [reply documentation/rfc/ptref_grammar.md]: reorganized

File: documentation/rfc/ptref_grammar.md (hunk 11 above)
Comment (TeXitoi) [reply documentation/rfc/ptref_grammar.md]: I prefer to olny have every collection or explicit collection, that's simpler to explain. span row need html by hand, and I'm lazy: https://stackoverflow.com/questions/46621765/can-i-merge-table-rows-in-markdown

**Review by pbougue — APPROVED** [review]

File: documentation/rfc/ptref_grammar.md (hunk 11 above)
Comment (TeXitoi) [reply documentation/rfc/ptref_grammar.md]: Compressed with "as above"

File: documentation/rfc/ptref_grammar.md [hunk 21]
```diff
@@ -0,0 +1,225 @@
...
+|`pred1 or pred2 - pred3 and pred4 or pred5`|`(pred1 or ((pred2 - pred3) and pred4)) or pred5`|
+ 
+ Examples:
+ * `/v1/coverage/id/lines?filter=physical_mode.id=physical_mode:Metro - stop_area.name=Nation - stop_area.name="Charles de Gaulle Etoile"` returns the metro lines that don't stop at Nation or Charles de Gaulle Etoile. Can also be written `/v1/coverage/id/lines?filter=physical_mode.id=physical_mode:Metro - (stop_area.name=Nation or stop_area.name="Charles de Gaulle Etoile")`
+ * `/v1/coverage/id/lines?filter=stop_area.name=Nation and stop_area.name="Charles de Gaulle Etoile" - (stop_area.name=Bastille or stop_area.name=Corvisart)` returns the lines that stop at Nation and Charles de Gaulle Etoile but not at Bastille nor Corvisart.
+
+## Sub request
+
+Method predicates allow you to request for a set of object of a given type, and then getting the corresponding objects of another type. Sub request allow you to chain such a pattern. Imagine you want to list the line of the same network of a given line. First, you want to get the networks corresponding to the given line, and then getting the corresponding lines from these networks.
```
Comment (stifoon) [inline documentation/rfc/ptref_grammar.md]: allowS

**Review by stifoon — COMMENTED** [review]
Master documentation for a main feature !

**Review by kinnou02 — APPROVED** [review]

**Pull request merged**